### PR TITLE
Update coreml codebook

### PR DESCRIPTION
**Pull request by metascroy**
This PR updates the dequantize_codebook quant primitive to be more compatible with CoreML.  More specifically:

* code_dtype is changed to nbits because CoreML cannot process a function that has non-standard dtypes in the signature (e.g., torch.uint3)

* This changes the codebook rank to codes.dim() + 2 to follow the convention here https://apple.github.io/coremltools/source/coremltools.converters.mil.mil.ops.defs.html#coremltools.converters.mil.mil.ops.defs.iOS18.compression.constexpr_lut_to_dense

### CODE REVIEW

File: torchao/prototype/quantization/codebook_coreml/codebook_quantized_tensor.py [hunk 1]
```diff
@@ -174,6 +177,17 @@ def _(func, types, args, kwargs):
...
+    assert len(args) == 2
+    indices, weight_tensor = (
+        args[0],
+        args[1],
+    )
+    weight_tensor = weight_tensor.dequantize()
+    return func(indices, weight_tensor, **kwargs)
+
+
 @implements([aten.detach.default, aten.alias.default])
```
Comment (jerryzh168) [inline torchao/prototype/quantization/codebook_coreml/codebook_quantized_tensor.py:191]: nit: can be a separate PR, you can try defining `CodebookTensor.tensor_data_names` and `CodebookTensor.tensor_attribute_names` removing these things now and see if it still works

https://github.com/pytorch/ao/pull/2597 and https://github.com/pytorch/ao/pull/2598 added more utils for TorchAOBaseTensor

File: torchao/prototype/quantization/codebook_coreml/codebook_ops.py [hunk 2]
```diff
@@ -48,43 +49,45 @@ def choose_qparams_and_quantize_codebook_coreml(
 
     Returns:
         Tuple[torch.Tensor, torch.Tensor]  The codebook (lookup table) Tensor and the quantized Tensor (codes, torch.uint8)
+        The LUT table has dimension input_tensor.dim() + 2, where:
```
Comment (jerryzh168) [inline torchao/prototype/quantization/codebook_coreml/codebook_ops.py]: nit: maybe spell out the dimensions with variables/expressions?

File: torchao/prototype/quantization/codebook_coreml/codebook_ops.py [hunk 3]
```diff
@@ -140,37 +150,54 @@ def dequantize_codebook(
...
-    original_shape = codes.shape
+    # Check that last two dimensions of codebook are [quant_levels, vec_dim]
+    assert codebook_shape[-2] == quant_levels, "Codebook shape mismatch with nbits"
 
-    # reshape to (N, num_lut, group_size)
-    codes = codes.reshape(codes.shape[0], num_lut, group_size)
-    dequant = torch.zeros_like(codes, dtype=output_dtype)
+    # Compute shape of lookup group indices from codes shape and block size
+    code_shape = codes.shape
+    ndim = len(code_shape)
```
Comment (jerryzh168) [inline torchao/prototype/quantization/codebook_coreml/codebook_ops.py]: nit: can do `codes.ndim`

File: torchao/prototype/quantization/codebook_coreml/codebook_ops.py [hunk 4]
```diff
@@ -140,37 +150,54 @@ def dequantize_codebook(
...
-    # dequant shape: (N, num_lut, group_size)
-    # codebook shape: (num_lut, 2 ** nbits)
-    # codes shape: (N, num_lut, group_size)
-    for i in range(num_lut):
-        # dequant[:, i, :]: (N, group_size)
-        # using squeeze to remove the training dim 1s after the lookup
-        dequant[:, i, :] = codebook[i][codes[:, i, :]].squeeze()
+    # Compute which codebook slice to use for each element
+    group_indices = []
+    for dim, bsz in zip(code_shape, block_size):
```
Comment (jerryzh168) [inline torchao/prototype/quantization/codebook_coreml/codebook_ops.py]: nit: dim might be confusing? this seems to be `code_size_i` and `block_size_i`

File: torchao/prototype/quantization/codebook_coreml/codebook_ops.py [hunk 5]
```diff
@@ -140,37 +150,54 @@ def dequantize_codebook(
...
+    mesh = torch.meshgrid(*group_indices, indexing="ij")
+    group_index_tensor = torch.stack(mesh, dim=-1)  # shape (..., N), where N = ndim
+
+    # Flatten everything to index efficiently
+    flat_codes = codes.reshape(-1)
+    flat_groups = group_index_tensor.reshape(-1, ndim)  # (..., ndim)
+
+    # Compute dequantized values via indexing
+    # index into codebook with (*group_index, code_index, :)
+    gathered = codebook[(*flat_groups.T, flat_codes)]  # shape (numel, vec_dim)
```
Comment (jerryzh168) [inline torchao/prototype/quantization/codebook_coreml/codebook_ops.py]: nit: can you add comments of some examples for these to make it easier to understand

File: torchao/prototype/quantization/codebook_coreml/codebook_quantized_tensor.py (hunk 1 above)
Comment (metascroy) [reply torchao/prototype/quantization/codebook_coreml/codebook_quantized_tensor.py:191]: This I will save for a future PR
